### PR TITLE
Findsymbols

### PR DESCRIPTION
**Pull request by nosami**
Adds /findsymbols endpoint to return all symbols in the workspace.

### CODE REVIEW

**Comment by aeschli** [comment]
@nosami - Can we add a search pattern as input, to limit the number of symbols returned? I'd suggest to do it similar to the CompletionText for AutoComplete (prefix match, substring match and camelcase).

**Comment by nosami** [comment]
@jrieken I never saw that before. Can you send me a link? I'll use that if it works.

I just had a look through here https://github.com/dotnet/roslyn/tree/b6484300dfafb43af0c27e542ec457a7583e1aa8/src/Workspaces/Core/Portable/FindSymbols but didn't see anything.

**Comment by nosami** [comment]
@aeschli sure!

**Comment by nosami** [comment]
I just found this internal method https://github.com/dotnet/roslyn/blob/b6484300dfafb43af0c27e542ec457a7583e1aa8/src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs#L160

Is that the one?

**Comment by jrieken** [comment]
@nosami - I was thinking of this one: https://github.com/dotnet/roslyn/blob/b6484300dfafb43af0c27e542ec457a7583e1aa8/src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_Declarations.cs#L260

**Comment by nosami** [comment]
Thanks! Will try it out tonight.

**Comment by nosami** [comment]
The SymbolFinder api was indeed a good choice :) Works well.

I added a new /findsymbolswithfilter endpoint. Not sure how you guys feel about this, but I couldn't find a way to make this backwards compatible (nullable request doesn't help) - maybe I'm just being dense here.

It now also returns types as well as methods etc. I'm thinking of leaving it this way and not bothering with a separate endpoint for types. What do you thing?

**Comment by nosami** [comment]
Oh... and as a side effect of doing this, I think I've also fixed the intellisense bug..... hit it when testing this code.

**Comment by jrieken** [comment]
Kudus for fixing the IntelliSense bug. Regarding ```/findSymbolsWithFilter``` it don't really mind, tho I would prefer a specialized request. Isn't [FromBody] smart enough to create an empty object (all members null) when there is no POST body at all?

**Comment by nosami** [comment]
No. Well, it didn't seem that way anyway. It was returning a 404 from the existing editor code when I had the request parameter in there. 

I tried `string filter` too instead of a request object. Same.

**Comment by jrieken** [comment]
I guess you already tried ```MyRequest? req=null```. Pretty sure there must be a way to achieve this? Right @davidfowl? Otherwise I am OK with another route

**Comment by nosami** [comment]
No... I forgot about that :S Let me try quickly.

**Comment by nosami** [comment]
Ok... That works :D Gimme a minute :)

File: src/OmniSharp/Api/Navigation/OmnisharpController.FindSymbols.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+        private QuickFix ConvertSymbol(ISymbol symbol, Location location)
+        {
+            var lineSpan = location.GetLineSpan();
+            var path = lineSpan.Path;
+            var documents = _workspace.GetDocuments(path);
+
+            return new QuickFix
+            {
+                Text = new SnippetGenerator().GenerateSnippet(symbol),
```
Comment (jrieken) [inline src/OmniSharp/Api/Navigation/OmnisharpController.FindSymbols.cs]: does this mean we have the snippet placeholder in the text?

Comment (nosami) [inline src/OmniSharp/Api/Navigation/OmnisharpController.FindSymbols.cs]: No. See the test output here https://github.com/OmniSharp/omnisharp-roslyn/blob/findsymbols/tests/OmniSharp.Tests/FindSymbolsFacts.cs#L44

I actually thought about renaming this class when I wrote this - the SnippetGenerator has 2 purposes :)

**Comment by jrieken** [comment]
:+1:

**Pull request merged**